src/main/cpp/win.cpp

```diff
@@ -601,6 +614,140 @@ Java_net_rubygrapefruit_platform_internal_jni_WindowsFileFunctions_readdir(JNIEn
     FindClose(dirHandle);
 }
 
+//
+// Returns "true" is the various fastReaddirXxx calls are supported on this platform.
+//
+JNIEXPORT jboolean JNICALL
+Java_net_rubygrapefruit_platform_internal_jni_WindowsFileFunctions_fastReaddirIsSupported(JNIEnv *env, jclass target) {
+#ifdef WINDOWS_MIN
+    return JNI_FALSE;
+#else
+    return JNI_TRUE;
+#endif
+}
+
+#ifndef WINDOWS_MIN
+typedef struct fast_readdir_handle {
+    HANDLE handle;
+    wchar_t* pathStr;
+} readdir_fast_handle_t;
+#endif
+
+#ifndef WINDOWS_MIN
+NTSTATUS invokeNtQueryDirectoryFile(HANDLE handle, BYTE* buffer, ULONG bufferSize) {
+    IO_STATUS_BLOCK ioStatusBlock;
+
+    return NtQueryDirectoryFile(
+        handle, // FileHandle
+        NULL, // Event
+        NULL, // ApcRoutine
+        NULL, // ApcContext
+        &ioStatusBlock, // IoStatusBlock
+        buffer, // FileInformation
+        bufferSize, // Length
+        FileIdFullDirectoryInformation, // FileInformationClass
+        FALSE, // ReturnSingleEntry
+        NULL, // FileName
+        FALSE); // RestartScan
+}
+#endif
+
+//
+// Returns a DirectByteBuffer pointing to a |fast_readdir_handle| structure on success
+// Returns NULL on failure (and sets error message in |result|).
+//
+JNIEXPORT jlong JNICALL
+Java_net_rubygrapefruit_platform_internal_jni_WindowsFileFunctions_fastReaddirOpen(JNIEnv *env, jclass target, jstring path, jobject result) {
+#ifdef WINDOWS_MIN
+    mark_failed_with_code(env, "Operation not supported", ERROR_CALL_NOT_IMPLEMENTED, NULL, result);
+    return NULL;
+#else
+    // Open file for directory listing
+    wchar_t* pathStr = java_to_wchar_path(env, path, result);
+    if (pathStr == NULL) {

```

Comment:
If there was an out of memory scenario, the code will crash before reaching this condition. It's most likely this condition is useless.

Reply:
I have no problem removing these checks, it is just that it is good practice to check for out of memory conditions. It could be, for example, that the String passed in from Java code is big (let's say 1MB), so converting it to UTF-16 may fail because allocating another 1MB fails. It does not necessarily mean the process/machine is out of memory.

That said, all other calls don't check the return value of "malloc", so I can remove these checks if you still think they are useless.

Reply:
You are right allocation should be checked. As you noted, there is no check in the original question. It may just get optimized out too.